### PR TITLE
Adds ability to modify the django runner class

### PR DESCRIPTION
**Pull request by kingbuzzman**
Addresses issue: https://github.com/behave/behave-django/issues/122
Replaces: https://github.com/behave/behave-django/pull/125

### CODE REVIEW

**Comment by kingbuzzman** [comment]
@bittner Still have 2 outstanding issues:
* the CI in `behave-django` fails to run.
* `bahave` hasn't published 1.2.7

**Comment by bittner** [comment]
Don't worry about Travis CI. I'll replace that by GitHub Actions tonight.

As for the Behave release, we have a much bigger problem: Only @jenisys can issue releases to PyPI, and it's unclear why he's waiting so long for it. (In the past this was all about Windows compatibility, but now ... not sure.)

Maybe you could (try to detect the Behave version and) abort execution when the feature is not supported, because the installed Behave package is not new enough?

**Comment by bittner** [comment]
The migration from Travis CI to GitHub Actions is now complete. I also added `isort` as a pipeline job to enforce more code style. The changes merge into the `main` branch (yes, `main` not `master`! I also switched the default branch).

You need to fix a minor conflict in the behave management command module now due to changes on the imports, I suppose. Sorry for the inconvenience! – Looking forward to a rebased PR that I could merge. :eyes: :100:

**Comment by kingbuzzman** [comment]
@bittner looks like my workflow is stuck in limbo awaiting your approval ;)

File: docs/conf.py
Comment (kingbuzzman) [inline docs/conf.py:17]: I didn't want to touch this, `isort` was complaining about it, and wouldn't shut up.

File: setup.py
Comment (kingbuzzman) [inline setup.py:7]: I didn't want to touch this, `isort` was complaining about it, and wouldn't shut up.

**Comment by kingbuzzman** [comment]
@bittner I "fixed" the requirement issue, I still don't like that 1.2.7 hasn't been released, but at least `tox` doesn't have an issue with it anymore. Don't know what you want/can to do about it... The remaining issue now is the CI not running.

File: HISTORY.rst
Comment (bittner) [inline HISTORY.rst:11]: How about
```suggestion
- New option to change the Django TestCase
```

**Review by bittner — APPROVED** [review]
LGTM, merging.

File: HISTORY.rst
Comment (kingbuzzman) [reply HISTORY.rst]: I like it

**Comment by kingbuzzman** [comment]
Well I would say it's "ready", my only concern is the [requirements.txt](https://github.com/behave/behave-django/blob/57f83817c9835cd1e25fc0cb76de3a9db080bf22/requirements.txt#L2), I suppose it's easy enough to create another PR to address it later on... balls on your court.

**Comment by bittner** [comment]
Ooops! What seems to be missing, actually, is updating the documentation. Can you add the option to the [configuration](docs/configuration.rst) chapter? With a few words of which problem it solves and how it is meant to be used.

Thank you!

**Comment by kingbuzzman** [comment]
> What seems to be missing, actually, is updating the documentation. Can you add the option to the configuration chapter? With a few words of which problem it solves and how it is meant to be used.

on it

> In the end, would you mind squashing all the changes, so that I can merge one (or two) pristine commit(s) bearing your name? Or should I simply let GitHub do the squashing?

I like the squashing option better.

**Comment by bittner** [comment]
> Well I would say it's "ready", my only concern is the [requirements.txt](https://github.com/behave/behave-django/blob/57f83817c9835cd1e25fc0cb76de3a9db080bf22/requirements.txt#L2), I suppose it's easy enough to create another PR to address it later on... balls on your court, chief ;) .

I'll take the opportunity to go down the stony way of trying to convince @jenisys to push a new release to PyPI. This is dangerous as it could keep this project trapped in the state that Behave is in for years now. But, in the end, _behave-django_ is just a development tool, isn't it? We can easily leave the dependency on the development version of Behave. As long as we want.

What do you think? Any other thoughts?

**Comment by kingbuzzman** [comment]
> > Well I would say it's "ready", my only concern is the [requirements.txt](https://github.com/behave/behave-django/blob/57f83817c9835cd1e25fc0cb76de3a9db080bf22/requirements.txt#L2), I suppose it's easy enough to create another PR to address it later on... balls on your court, chief ;) .
> 
> I'll take the opportunity to go down the stony way of trying to convince @jenisys to push a new release to PyPI. This is dangerous as it could keep this project trapped in the state that Behave is in for years now. But, in the end, _behave-django_ is just a development tool, isn't it? We can easily leave the dependency on the development version of Behave. As long as we want.
> 
> What do you think? Any other thoughts?

I agree with everything you just said 👍

**Comment by kingbuzzman** [comment]
@bittner ..why do you have to accept every CI commit now? Is there a way you can set it to run always? Are you charged per CI task?

**Comment by bittner** [comment]
> @bittner ..why do you have to accept every CI commit now? Is there a way you can set it to run always? Are you charged per CI task?

I think this is just as long as I haven't merged "your first contribution". This sounds silly, because you have contributed a lot, I feel, be GitHub has no evidence (no commit from you in the code base) yet. Don't worry, we're about to fix that, too.

File: behave_django/management/commands/behave.py
Comment (bittner) [inline behave_django/management/commands/behave.py:64]: It's a bit sad that we can't simply set `behave_django.runner.BehaviorDrivenTestRunner` as a `default` attribute in this parser argument (because we have it [hard-coded in the management command](https://github.com/behave/behave-django/blob/main/behave_django/management/commands/behave.py#L144)). I feel that would make things very elegant and transparent. :worried:

Comment (kingbuzzman) [reply behave_django/management/commands/behave.py:64]: I played with this exact idea, and turned it down, because of how the (`simple`/`use-existing`) flags work. It could be worth exploring in the future to simply disable those flags and explicitly state you want them by doing: `--runner-class behave_django.runner.Simple.....`

But that can be down the line

File: behave_django/management/commands/behave.py
Comment (bittner) [inline behave_django/management/commands/behave.py:157]: Can't we simply use this line in the `else` clause and have `BehaviorDrivenTestRunner` set as a default in the `--runner-class` parser argument? :bulb:

File: behave_django/management/commands/behave.py
Comment (bittner) [inline behave_django/management/commands/behave.py:141]: We could check for a deviation from the default value instead of just `options['runner_class']`, e.g.

```suggestion
        if options['runner_class'] != DEFAULT_RUNNER and both_active:
```

Then again, do the switches `--use-existing-database` and `--simple` (always or at all) make sense with a custom runner? We might make those switches mutually exclusive, otherwise. Would that be a sane compromise?

Comment (kingbuzzman) [reply behave_django/management/commands/behave.py:141]: Just committed, see what you think

Comment (kingbuzzman) [reply behave_django/management/commands/behave.py:141]: Yup. On it ;)

File: HISTORY.rst
Comment (bittner) [inline HISTORY.rst:11]: Isn't this really an option to change the test runner? We don't supply a custom [test case class](https://github.com/behave/behave-django/blob/main/behave_django/testcase.py) with the new option, do we?

Comment (kingbuzzman) [reply HISTORY.rst:11]: Crud. I got my test classes mixed up, yes, these are TestCase, not TestRunner. I'll change the names now.

Comment (kingbuzzman) [reply HISTORY.rst]: Ok, can you let this last commit run tests, let's see if it passes *fingers crossed*

Comment (bittner) [reply HISTORY.rst]: Alright. The pipeline passes. Let's merge the changes?

Comment (kingbuzzman) [reply HISTORY.rst]: Push it real good ;)

File: docs/configuration.rst
Comment (bittner) [inline docs/configuration.rst:39]: It might make sense to [directly link](https://behave.readthedocs.io/en/latest/behave.html#cmdoption-runner-class) to Behave's documentation, what do you think?

File: docs/configuration.rst
Comment (kingbuzzman) [reply docs/configuration.rst]: ?

Comment (bittner) [reply docs/configuration.rst]: Yes, sure. Let's clean up the option list in a proper manner. :+1:

Comment (kingbuzzman) [reply docs/configuration.rst]: done

File: docs/configuration.rst
Comment (bittner) [reply docs/configuration.rst:72]: You could use (anonymous) in-line links like this:

```reStructuredText
The truth is in the `behave docs <https://example.com>`__ over there.
```

Two underscores make the hyper-reference anonymous, a single one makes it reusable.

But there should be a way with reusing an existing reference or link... :thinking:

Comment (bittner) [reply docs/configuration.rst:72]: Or do you want this?

```reStructuredText
in the `behave docs`_.

.. _behave docs: https://behave.readthedocs.io/en/latest/behave.html#cmdoption-runner-class
```

Comment (bittner) [reply docs/configuration.rst:72]: BTW, if you want easier communication: Want to meet on Gitter? https://gitter.im/behave/behave-django

Comment (kingbuzzman) [reply docs/configuration.rst]: I think what i did "works", for now. Let me know.

**Comment by kingbuzzman** [comment]
Very nice. Thank you @bittner

**Comment by bittner** [comment]
Thank you so much, @kingbuzzman! :1st_place_medal: :100: :trophy: 

As a bonus you get free, automatic CI runs for all your changes with your next PRs!! :smirk:

**Comment by bittner** [comment]
BTW, have you seen #129? Maybe you want to "apply"? :nerd_face: :smiley: 

I would appreciate that. Dead seriously.

**Comment by kingbuzzman** [comment]
I did see it, I wouldn't mind as a soft/light co-maintainer

**Pull request merged**
